src/Compilers/Core/Portable/SourceGeneration/Nodes/CombineNode.cs

```diff
@@ -44,29 +52,50 @@ public NodeStateTable<(TInput1, TInput2)> UpdateStateTable(DriverStateTable.Buil
 
             // get the input2 item
             var isInput2Cached = input2Table.IsCached;
-            TInput2 input2 = input2Table.Single();
+            (TInput2 input2, IncrementalGeneratorRunStep? input2Step) = input2Table.Single();
 
             // append the input2 item to each item in input1 
             foreach (var entry1 in input1Table)
             {
-                var state = (entry1.state, isInput2Cached) switch
+                var stopwatch = SharedStopwatch.StartNew();
+
+                var stepInputs = builder.TrackIncrementalSteps ? ImmutableArray.Create((entry1.Step!, entry1.OutputIndex), (input2Step!, 0)) : default;
+
+                var state = (entry1.State, isInput2Cached) switch
                 {
                     (EntryState.Cached, true) => EntryState.Cached,
                     (EntryState.Cached, false) => EntryState.Modified,
-                    _ => entry1.state
+                    _ => entry1.State
                 };
 
-                var entry = (entry1.item, input2);
-                if (state != EntryState.Modified || _comparer is null || !builder.TryModifyEntry(entry, _comparer))
+                var entry = (entry1.Item, input2);
+                if (state != EntryState.Modified || _comparer is null || !builder.TryModifyEntry(entry, _comparer, stopwatch.Elapsed, stepInputs, state))
                 {
-                    builder.AddEntry(entry, state);
+                    builder.AddEntry(entry, state, stopwatch.Elapsed, stepInputs, state);
                 }
             }
 
             return builder.ToImmutableAndFree();
         }
 
-        public IIncrementalGeneratorNode<(TInput1, TInput2)> WithComparer(IEqualityComparer<(TInput1, TInput2)> comparer) => new CombineNode<TInput1, TInput2>(_input1, _input2, comparer);
+        private NodeStateTable<(TInput1, TInput2)> RecordStepsForCachedTable(DriverStateTable.Builder graphState, NodeStateTable<(TInput1, TInput2)> previousTable, NodeStateTable<TInput1> input1Table, NodeStateTable<TInput2> input2Table)
+        {
+            Debug.Assert(input1Table.HasTrackedSteps && input2Table.IsCached);
+            var builder = graphState.CreateTableBuilder(previousTable, _name);
+            (_, IncrementalGeneratorRunStep? input2Step) = input2Table.Single();
+            foreach (var entry in input1Table)
+            {
+                var stepInputs = builder.TrackIncrementalSteps ? ImmutableArray.Create((entry.Step!, entry.OutputIndex), (input2Step!, 0)) : default;
```

Comment:
>builder.TrackIncrementalSteps [](http://example.com/codeflow?start=33&length=29)

The caller is already checking `graphState.DriverState.TrackIncrementalSteps` so will this ever be `false`?